Fix attribute ManyToMany mapping

BoShurik opened the pull request:
Fixes #9902

greg0ire commented:
Can you please write a test for this, based on one of the examples mentioned in the issue you fix?

BoShurik commented:
@greg0ire done

greg0ire commented on tests/Doctrine/Tests/ORM/Mapping/AttributeDriverTest.php:
```diff
@@ -123,6 +124,45 @@ public function testLegacyInheritance(): void
+    {
+        if (PHP_VERSION_ID < 80100) {
+            self::markTestSkipped('Does not work on PHP 8.0.* due to nested attributes missing.');
+        }
```
You can use `@requires PHP 8.1` for this.

greg0ire commented on tests/Doctrine/Tests/ORM/Mapping/Fixtures/AttributeEntityWithNestedJoinColumns.php:
```diff
@@ -0,0 +1,26 @@
+    #[ORM\GeneratedValue]
+    public $id;
+
+    /** @var AttributeEntityWithNestedJoinColumns[] */
```
This is not supposed to be an array, is it?

```suggestion
    /** @var Collection<AttributeEntityWithNestedJoinColumns> */
```

BoShurik replied:
Copy pasted from https://github.com/doctrine/orm/blob/2.15.x/tests/Doctrine/Tests/ORM/Mapping/AttributeDriverTest.php#L139. Does it make sense to change that there too?

greg0ire replied:
It would I suppose, but in a separate commit.

greg0ire reviewed (commented):
I've tested locally, and the both assertions fail without your changes 👍

Please improve your commit message [according to our contributing guide](https://www.doctrine-project.org/contribute/index.html)

Right now, one has to read the code and tests to realize what's broken and how you fix it.

greg0ire commented on lib/Doctrine/ORM/Mapping/Driver/AttributeDriver.php:
```diff
@@ -433,6 +433,14 @@ public function loadMetadataForClass($className, PersistenceClassMetadata $metad
+
+                    foreach ($joinTableAttribute->inverseJoinColumns as $joinColumn) {
+                        $joinTable['inverseJoinColumns'][] = $this->joinColumnToArray($joinColumn);
+                    }
```
I've looked into the issue, and it looks like there is a piece of code that is just missing from this new implementation. I don't know why it was done like that, but your commit message could look like this:

```
Take into account join columns specifications.

Currently, the AttributeDriver ignores any join column attribute
specified on a many to many relationship.
Let's copy code from the AnnotationDriver to fix that.

Fixes #9902
```

greg0ire replied:
It looks like some other stuff was forgotten in #8266 in this code block: d7d6b9d2c7fcfc16a6a706c78161b441aed39ffb

BoShurik replied:
It was missing because php 8.0 does not support nested attributes

greg0ire replied:
Ah great! That explains it fully!

BoShurik committed:
Update doc-block for AttributeDriverTest's fixtures

BoShurik committed:
Take into account join columns specifications.

Currently, the AttributeDriver ignores any join column attribute specified on a many to many relationship.
Let's copy code from the AnnotationDriver to fix that.

Fixes #9902

greg0ire commented:
Thanks @BoShurik !
